### DISCUSSION TITLE
How to write a log file in text mode (rather than in binary mode)?

### DISCUSSION THREAD

**Discussion by N-Dekker** [gabime/spdlog#2634] [Q&A] [answered]
When a file is specified as sink, it appears to be opened in binary mode: the letter `b` in both `"ab"` and `"wb"`, at https://github.com/gabime/spdlog/blob/927cc29444a294d76e83dfb898e797dc431ce094/include/spdlog/details/file_helper-inl.h#L37-L38

As a consequence, a `\n` with a string passed to the logger won't be converted automatically to `\r\n`, even though this might be preferable on Windows. For example: 

    auto text_file_sink = std::make_shared<spdlog::sinks::basic_file_sink_mt>("my-log.txt");
    text_file_sink->set_pattern("%v");
    spdlog::default_logger()->sinks().push_back(text_file_sink);
    spdlog::info("Line one\nLine two");

The resulting text file ("my-log.txt") looks as follows:

    Line one\n
    Line two\r\n

Is it be possible to tweak spdlog to write a log file in text mode, so that on Windows, each `\n` would be translated automatically to `\r\n`?

**Comment by tt4g** [comment, marked as answer]
spdlog does not provide an option to change the file open mode.
You can open pull requests if you need such a feature.

It should be possible to add something like an `SPDLOG_EOL` macro that changes the newline character that spdlog adds to the end of a line.

https://github.com/gabime/spdlog/blob/927cc29444a294d76e83dfb898e797dc431ce094/include/spdlog/tweakme.h#L62-L66
https://github.com/gabime/spdlog/blob/654dbc5c3291a0981e6c473fac199409d41fcde8/include/spdlog/details/os.h#L23-L32

A workaround available in spdlog v1.11.0 is to create a custom formatter.
Wiki: https://github.com/gabime/spdlog/wiki/3.-Custom-formatting#extending-spdlog-with-your-own-flags

It should be possible to implement this with reference to `v_formatter` which handles the `%v` flag.
Just output an additional `\r` when `\n` appears in `msg.paylod`.

https://github.com/gabime/spdlog/blob/927cc29444a294d76e83dfb898e797dc431ce094/include/spdlog/pattern_formatter-inl.h#L679-L692

**Reply by N-Dekker** [reply]
Thanks @tt4g I'm preparing a pull request: https://github.com/N-Dekker/spdlog/tree/file_sink-text-mode (still under construction)

Update: the pull request is there, please check: https://github.com/gabime/spdlog/pull/2636
